[AArch64] Optimize CBZ wzr and friends.

davemgreen opened the pull request:
In certain situations, especially with zero phi operands propagated after tail
duplications, we can end up with CBZ/CBNZ/TBZ/TBNZ with a zero register. It
can can be introduced late in the pipeline.

This patch adds a basic pass to fold them away to either a direct branch or
removing the instruction entirely. It runs quite late n the pipeline, so doesnt
fit into any of the existing passes. It only needs to look at the terminators
to each BB, so the new pass should have a limited in compile-time impact.

dtellenbach commented:
Would it make sense to also check for `cb.{eq,ne} {x,w}zr, #0`? Not sure if we'd ever codegen that though.

davemgreen committed:
[AArch64] Optimize CBZ wzr and friends.

In certain situations, especially with zero phi operands propagated after tail
duplications, we can end up with CBZ/CBNZ/TBZ/TBNZ with a zero register. It
can can be introduced late in the pipeline.

This patch adds a basic pass to fold them away to either a direct branch or
removing the instruction entirely. It runs quite late n the pipeline, so doesnt
fit into any of the existing passes. It only needs to look at the terminators
to each BB, so the new pass should have a limited in compile-time impact.

dtellenbach commented on llvm/lib/Target/AArch64/AArch64RedundantCondBranchPass.cpp:
```diff
@@ -0,0 +1,113 @@
+          if (S != Target)
+            MBB->removeSuccessor(S);
+        SmallVector<MachineInstr *> DeadInstrs;
+        for (auto It = MI.getIterator(); It != MBB->end(); ++It)
```
Can you erase these right away here?

usha1830 commented on llvm/lib/Target/AArch64/AArch64RedundantCondBranchPass.cpp:
```diff
@@ -0,0 +1,113 @@
+//=- AArch64RedundantCondBranch.cpp - Remove redundant cbz wzr --------------=//
```
```suggestion
//=- AArch64RedundantCondBranch.cpp - Remove redundant conditional branches --------------=//
```

usha1830 commented on llvm/lib/Target/AArch64/AArch64RedundantCondBranchPass.cpp:
```diff
@@ -0,0 +1,113 @@
+    case AArch64::CBNZX:
+    case AArch64::TBNZW:
+    case AArch64::TBNZX:
+      // CBNZ XZR -> nop
```
```suggestion
      // CBNZ/TBNZ with WZR/XZR -> never taken, remove branch and successor
```

usha1830 reviewed (approved):
LGTM, Thanks!

davemgreen committed:
Address comments and cleanup

dtellenbach reviewed (approved):
LGTM, thanks!
